### PR TITLE
Blend layers using integer arithmetics.

### PR DESCRIPTION
**Pull request by wertarbyte**

### CODE REVIEW

File: src/background_remover.cc [hunk 1]
```diff
@@ -184,7 +184,7 @@ cv::Mat BackgroundRemover::getMaskFromOutput() {
 
         ret.forEach([&](unsigned char &is_background, const int loc[]) -> void {
             const auto p = cv::Point(loc[1], loc[0]);
-            is_background = maxIdx(labels(p)) != person_label;
+            is_background = 255 * (maxIdx(labels(p)) != person_label);
```
Comment (yath) [inline src/background_remover.cc]: is_background isn’t really appropriate now anymore. background_depth or so? (Also, please make this `maxIdx(..) == bar ? 0 : 255` or something more readable than a multiplication with a bool.)

File: src/background_remover.cc [hunk 2]
```diff
@@ -268,8 +268,38 @@ static cv::Mat resizeAndPadTo(const cv::Mat &frame, int targetw, int targeth, Pa
...
+        mix += (p * b[c]) / 255;
+        mix += ((255-p) * a[c]) / 255;
+        result[c] = mix;
+    }
+}
+
+void BackgroundRemover::blendLayers(const cv::Mat &foreground,
+                                    const cv::Mat &background,
+                                    const cv::Mat &mask,
+                                    cv::Mat &output) {
```
Comment (yath) [inline src/background_remover.cc]: Return that

File: src/background_remover.cc [hunk 3]
```diff
@@ -268,8 +268,38 @@ static cv::Mat resizeAndPadTo(const cv::Mat &frame, int targetw, int targeth, Pa
     return ret;
 }
 
-void BackgroundRemover::maskBackground(cv::Mat &frame /* rgb */, const cv::Mat &maskImage /* rgb */,
-                                       bool do_blur_mask, Timing &t) {
+void BackgroundRemover::blendPixels(const cv::Vec3b &a,
+                                    const cv::Vec3b &b,
+                                    uint8_t p,
+                                    cv::Vec3b &result) {
```
Comment (yath) [inline src/background_remover.cc]: Return that

File: src/background_remover.cc [hunk 4]
```diff
@@ -268,8 +268,38 @@ static cv::Mat resizeAndPadTo(const cv::Mat &frame, int targetw, int targeth, Pa
...
+    for (int c = 0; c < 3; c++) {
```
Comment (yath) [inline src/background_remover.cc]: You shouldn’t need to iterate here. Looking at opencv/modules/core/include/opencv2/core/matx.hpp, operator* (for example) is implemented as:

```
template<typename _Tp, int cn> static inline
Vec<_Tp, cn>& operator *= (Vec<_Tp, cn>& a, int alpha)
{
    for( int i = 0; i < cn; i++ )
        a[i] = saturate_cast<_Tp>(a[i]*alpha);
    return a;
}
```

File: src/background_remover.cc [hunk 5]
```diff
@@ -268,8 +268,38 @@ static cv::Mat resizeAndPadTo(const cv::Mat &frame, int targetw, int targeth, Pa
     return ret;
 }
 
-void BackgroundRemover::maskBackground(cv::Mat &frame /* rgb */, const cv::Mat &maskImage /* rgb */,
-                                       bool do_blur_mask, Timing &t) {
+void BackgroundRemover::blendPixels(const cv::Vec3b &a,
```
Comment (yath) [inline src/background_remover.cc]: This (and the following) function don’t seem to require access to `this`, so they could be declared static (not: a static member function). In fact, this being a method prevents the compiler from inlining this.

File: src/background_remover.cc [hunk 6]
```diff
@@ -268,8 +268,38 @@ static cv::Mat resizeAndPadTo(const cv::Mat &frame, int targetw, int targeth, Pa
...
+        mix += ((255-p) * a[c]) / 255;
+        result[c] = mix;
+    }
+}
+
+void BackgroundRemover::blendLayers(const cv::Mat &foreground,
+                                    const cv::Mat &background,
+                                    const cv::Mat &mask,
+                                    cv::Mat &output) {
+    for (int x = 0; x < foreground.cols; x++) {
```
Comment (yath) [inline src/background_remover.cc]: Please use `ret.forEach`. (Please name it `ret` for consistency too.)

File: src/background_remover.cc [hunk 7]
```diff
@@ -268,8 +268,38 @@ static cv::Mat resizeAndPadTo(const cv::Mat &frame, int targetw, int targeth, Pa
...
+        for (int y = 0; y < foreground.rows; y++) {
+            cv::Point p(x, y);
+            uint8_t m = mask.at<uint8_t>(p);
```
Comment (yath) [inline src/background_remover.cc]: Please use `unsigned char` for consistency like in the caller.

File: src/background_remover.cc [hunk 8]
```diff
@@ -306,7 +336,11 @@ void BackgroundRemover::maskBackground(cv::Mat &frame /* rgb */, const cv::Mat &
         cv::blur(mask, mask, cv::Size(blur_width, blur_width));
     }
 
-    maskImage.copyTo(frame, mask);
+    if (do_blend_layers) {
+      blendLayers(frame, maskImage, mask, frame);
```
Comment (yath) [inline src/background_remover.cc]: I’m assuming it’s cv::blur that sets pixel values to !255? If so, will `blendLayers` do anything different than `maskImage.copyTo` if `do_blur_mask` is false? Maybe add a warning that blending is only effective when also smoothing, or just move this `if` into the one above. (I assume `copyTo` is still quicker than `blendLayers` with a 255-or-0-weight.)

**Review by yath — CHANGES_REQUESTED** [review]
Cool, thanks!

File: src/background_remover.cc (hunk 4 above)
Comment (wertarbyte) [reply src/background_remover.cc]: As discussed on IRC, this seems to be the least messy way to do this :-D
I'll replace the hardcoded 3 with the .channels attribute of the vector...

File: src/background_remover.cc (hunk 7 above)
Comment (wertarbyte) [reply src/background_remover.cc]: Fine, but it's sooooooo long :-D

File: src/background_remover.cc (hunk 8 above)
Comment (wertarbyte) [reply src/background_remover.cc]: It does _something_, although I don't know exactly what or why.

Comment (wertarbyte) [reply src/background_remover.cc]: Scaling the mask already creates values in between, so there is some effect even without blurring.

**Pull request merged**